keps/sig-network/4631-loadbalancerstatus/README.md

```diff
@@ -0,0 +1,1066 @@
+  single feature (e.g., UDP, or source ranges support) so they could
+  be made to bail out early on unsupported clouds.)
+
+### Non-Goals
+
+- Requiring all cloud providers to implement all existing
+  load-balancer-related features.
+
+[cloud-provider-gcp]: https://github.com/kubernetes/cloud-provider-gcp/
+[cloud-provider-kind]: https://github.com/kubernetes-sigs/cloud-provider-kind/
+
+## Proposal
+
+1. Define new conditions for `Service.Status.Conditions` to allow the
+   cloud provider to communicate load balancer status better to the
+   user.
+
+2. Define clearer rules about what cloud providers are required to pay
+   attention to in the Service object, and how they should behave when
+   they can't provision a load balancer that exactly implements the
+   requested semantics.
+
+3. Define new API to allow the user to specify when they would prefer
+   that the cloud fail to provision a load balancer rather than
+   partially provisioning it.
+
+### Risks and Mitigations
+
+Using the new features should not introduce any new risks.
+
+They should hopefully _improve_ security for users. Currently, you
+sometimes get load balancers that aren't what you wanted (e.g.,
+failing to implement source ranges or session affinity). This KEP
+makes it possible for users to prevent this from happening.
+
+(If a cloud provider implemented the API incorrectly and claimed
+support for features that it didn't actually support, then that could
+lead to users being more confident when deploying broken load
+balancers than they are now, but the cloud provider would fail the e2e
+tests in that case, so this shouldn't happen.)
+
+## Design Details
+
+### `Service.Status.Conditions`
+
+The problem of indicating provisioning progress and provisioning
+failure can easily be solved with conditions on the Service.
+
+We will add three new Service conditions:
+
+  - `LoadBalancerServing`, to indicate explicitly that the load
+    balancer is (or isn't) serving traffic.
+
+  - `LoadBalancerProvisioning`, to indicate that the cloud provider
+    is (or isn't) configuring/provisioning the load balancer.
+
+  - `LoadBalancerDegraded`, to indicate that the load balancer is
+    aware that it is not fully implementing the semantics requested by
+    the user, in cases where implementing the full semantics is not
+    mandatory.
+
+#### The `LoadBalancerServing` Condition
+
+The `LoadBalancerServing` condition, when `True`, indicates that
+the service's load balancer is handling connections via the IP(s) or
+hostname(s) indicated in `.Status.LoadBalancer`.
+
+The cloud provider should set a `LoadBalancerServing` condition
+immediately upon first becoming aware of a `Type: LoadBalancer`
+Service. At all times, the condition should be `True` if the load
+balancer is handling connections, and `False` if it is not. (There are
+no defined semantics for an `Unknown` value of this condition.)
+
+(This means that at construction time, the `LoadBalancerServing`
```

Comment:
Pedantically, the `.Status.LoadBalancer` field means "I have an IP allocated".  It doesn't mean that traffic is flowing or that the IP is reachable yet.

For example, it may take seconds or even minutes for infrastructure to actually start receiving and routing traffic.

Does "LoadBalancerServing" mean that we have provisioned the LB or that someone has confirmed that the traffic is flowing (from where?  what about LB source ranges?)  

Does this line up with Gateway's conditions?

Reply:
Somewhere else I claim that the e2e tests will expect to be able to connect immediately upon `LoadBalancerServing` becoming `True` (though we could change that).

I think my intuition was "Serving means that the cloud provider _believes_ that the LB is accepting/redirecting connections". It definitely doesn't imply that the cloud provider has actually tried making a connection. (In addition to source ranges, there's the problem that maybe the cluster end of the service is currently not-ready, which shouldn't affect the LB condition.)

If there is internal status information that tells the cloud provider that the LB has been allocated but is not yet actually serving, then the cloud provider should not mark the service as LoadBalancerServing until the internal status information says it's serving.

If there's an allocated-but-not-yet-serving state, but the cloud provider _doesn't_ have any way of figuring that out other than by trying to connect to the LB, then I guess it needs to call that state "serving"...